tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,197 @@
+# Lint as: python2, python3
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TFX BeamFusionOptimizer definition."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from typing import List
+
+from tfx.components.base import base_node
+from tfx.components.base import base_executor
+from tfx.orchestration.pipeline import Pipeline
+
+class BeamFusionOptimizer(object):
+  """Optimizer for TFX pipelines, utilizing Beam Fusion"""
+
+  def __init__(self, pipeline: Pipeline):
+    self.pipeline = pipeline
+
+  def _topologically_sort(self, components: List[base_node.BaseNode],
```

Comment:
Not sure this is correct. Consider:
```
A
|
C
|
D  B
| /
E
```
where the roots are `[A, B]`. Then the layers will be `[[A, B], [C, E], [D]]` where `E` is visited before `D`.

Reply:
Nice catch! I rewrote the `_topologically_sort()` method to account for multiple sources. I actually had modeled it off of the topological sort in `pipeline.py` seen here https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/pipeline.py#L174, maybe that logic should also be rewritten.